applications/serial_lte_modem/src/slm_cmux.c

```diff
@@ -0,0 +1,187 @@
+ */
+#include "slm_cmux.h"
+#include "slm_at_host.h"
+#include "slm_uart_handler.h"
+#include <zephyr/logging/log.h>
+#include <zephyr/modem/backend/uart.h>
+#include <zephyr/modem/cmux.h>
+#include <zephyr/modem/pipe.h>
+#include <assert.h>
+
+LOG_MODULE_REGISTER(slm_cmux, CONFIG_SLM_LOG_LEVEL);
+
+/* This makes use of some of the Zephyr modem subsystem that has a CMUX module implementation.
+ */
+
+static struct {
+	/* UART backend */
+	struct modem_pipe *uart_pipe;
+	struct modem_backend_uart uart_backend;
+	uint8_t uart_backend_receive_buf[512];
+	uint8_t uart_backend_transmit_buf[512];
+
+	/* CMUX */
+	struct modem_cmux instance;
+	uint8_t cmux_receive_buf[128];
+	uint8_t cmux_transmit_buf[256];
+
+	/* CMUX channels (Data Link Connection Identifier); index = address - 1 */
+	struct cmux_dlci {
+		struct modem_cmux_dlci instance;
+		struct modem_pipe *pipe;
+		uint8_t *recv_buf;
+		uint16_t recv_buf_size;
+		uint8_t address;
+	} dlcis[2];
+} cmux;
+
+static void dlci_pipe_event_handler(struct modem_pipe *pipe,
+				    enum modem_pipe_event event, void *user_data)
+{
+	const struct cmux_dlci *const dlci = user_data;
+
+	switch (event) {
+	case MODEM_PIPE_EVENT_OPENED:
+	case MODEM_PIPE_EVENT_CLOSED:
+		LOG_INF("DLCI %u %s.", dlci->address,
+			(event == MODEM_PIPE_EVENT_OPENED) ? "opened" : "closed");
+		break;
+	case MODEM_PIPE_EVENT_RECEIVE_READY:
+		static uint8_t recv_buf[256];
+		const int recv_len = modem_pipe_receive(pipe, recv_buf, sizeof(recv_buf));
+
+		if (recv_len > 0) {
+			LOG_DBG("Received %u bytes on DLCI %u.", recv_len, dlci->address);
+			modem_pipe_transmit(pipe, recv_buf, recv_len);
+		} else {
+			LOG_ERR("Failed to receive data from DLCI %u (%d).",
+				dlci->address, recv_len);
+		}
+		break;
+	}
+}
+
+static void cmux_event_handler(struct modem_cmux *, enum modem_cmux_event event, void *)
+{
+	assert(event == MODEM_CMUX_EVENT_CONNECTED || event == MODEM_CMUX_EVENT_DISCONNECTED);
+	LOG_INF("CMUX %sconnected.", (event == MODEM_CMUX_EVENT_CONNECTED) ? "" : "dis");
+}
+
+static void init_dlci(size_t address, uint16_t recv_buf_size,
+		      uint8_t recv_buf[static recv_buf_size])
+{
+	assert(ARRAY_SIZE(cmux.dlcis) >= address);
+
+	struct cmux_dlci *const dlci = &cmux.dlcis[address - 1];
+	const struct modem_cmux_dlci_config dlci_config = {
+		.dlci_address = address,
+		.receive_buf = recv_buf,
+		.receive_buf_size = recv_buf_size
+	};
+
+	dlci->pipe = modem_cmux_dlci_init(&cmux.instance, &dlci->instance, &dlci_config);
+	dlci->recv_buf = recv_buf;
+	dlci->recv_buf_size = recv_buf_size;
+	dlci->address = address;
+
+	modem_pipe_attach(dlci->pipe, dlci_pipe_event_handler, dlci);
+}
+
+static int cmux_start(void)
+{
+	int ret;
+
+	/* Disable SLM UART so that CMUX can take over it. */
+	ret = slm_uart_rx_disable();
+	if (ret) {
+		return ret;
+	}
+
+	{
+		const struct modem_backend_uart_config uart_backend_config = {
+			.uart = DEVICE_DT_GET(DT_CHOSEN(ncs_slm_uart)),
+			.receive_buf = cmux.uart_backend_receive_buf,
```

Comment:
Does this use ISR or async uart? I am wondering as with async uart, there should be several buffers, or the performance is going to suffer drastically.

Reply:
Async. Internally ring buffers are used, so I imagine that it's equivalent to using several buffers?

Reply:
I am not sure what they have done internally, but the idea is to have several buffers so that uart receiving can continue while you process the other buffer. Otherwise this will always stop the uart from receiving when you are processing the buffer.

Reply:
I checked, and internally the async UART backend splits the supplied buffer into several sub-buffers (in 4; 2 buffers and 2 ring buffers).
Also, there is actually quite a bit of copying happening between the modules (which is why we have so many buffers declared), so I think that by the time data goes through a `modem_pipe` the end that transmitted it (e.g. the async UART backend when receiving) doesn't need to keep the data anymore even though it may not even have reached the application yet.